IPIP/0373-double-hash-dht.md

```diff
@@ -0,0 +1,404 @@
+- A DHT Server returns all of the Provider Records matching to at most **`MatchLimit = 64`** distinct `HASH2`. Magic number explanation in [k-anonymity](#k-anonymity).
+- Provider Record Timestamp (`TS`) validity period: `48h`
+
+### Definitions
+
+- **`CID`** is the IPFS [Content IDentifier](https://github.com/multiformats/cid)
+- **`MH`** is the [Multihash](https://github.com/multiformats/multihash) contained in a `CID`. It corresponds to the digest of a hash function over some content. `MH` is represented as a 32-byte array.
+- **`HASH2`** is defined as `SHA256(bytes("CR_DOUBLEHASH") || MH)`. It represents the location of the Kademlia keyspace for the Provider Record associated with `CID`. `HASH2` is represented as a 32-byte array. `HASH2 = SHA256(bytes("CR_DOUBLEHASH") || MH)`.
+- **Content Provider** is the node storing some content, and advertising it to the DHT.
+- **DHT Servers** are nodes running the IPFS public DHT. In this documents, DHT Servers mostly refer to the DHT Servers storing the Provider Records associated with specific `CID`s, and not the DHT Servers helping routing lookup requests to the right keyspace location. 
+- **Client** is an IPFS client looking up a content identified by an already known `CID`.
+- **Publish Process** is the process of the Content Provider communicating to the DHT Servers that it provides some content identified by `CID`.
+- **Lookup Process** is the process of the Client retrieving the content identified by `CID`.
+- **`PeerID`** s define stable [peer identities](https://github.com/libp2p/specs/blob/master/peer-ids/peer-ids.md). The `PeerID` is derived from the node's cryptographic public key.
+- **`multiaddrs`** are the [network addresses](https://github.com/libp2p/specs/tree/master/addressing) associated with a `PeerID`. It represents the location(s) of the peer.
+- **`KeyPrefix`** is defined as a prefix of length `l` bits of `HASH2`. `KeyPrefix` is represented by a `byte` concatenated with a variable sized array of bytes, containing at most 32 bytes. The leading `byte` represents the binary representation of `l - 1`, making prefixes of length `256` possible, but not prefixes of length `0`. The trailing byte array is of length `ceil(l/8)` bytes, and its content is the bits prefix right padded with zeros.
+- **`ServerKey`** is defined as `SHA256(bytes("CR_SERVERKEY") || MH)`. It is derived from the `MH`. The Content Provider communicates `ServerKey` to the DHT Servers during the Publish Process. The DHT Servers use `ServerKey` to encrypt `TS`, `Signature` and Content Providers `multiaddrs` sent to the Client when some Provider Records match the requested `Prefix`. `ServerKey` is represented as a 32-byte array.
+- **`TS`** is the [Unix Timestamp](https://en.wikipedia.org/wiki/Unix_time) corresponding content publish time. `TS` is represented as a 32-bit **unsigned** Integer, allowing timestamps to range from `1970-01-01T00:00:00Z` to `2106-02-07T06:28:15Z` before reaching the overflow.
```

Comment:
Initially it was to spare bytes on the wire. However, as we need a 12-byte IV for AES-GCM, we could define the IV to be 8 bytes Timestamp (publish time) + 4 random bytes. So we don't use extra capacity on the wire. WDYT @Jorropo ?

Reply:
1. IV should be unique for each encryption key (so unless you want to create billions of distinct provider records for the same content within the same second you should be safe)
2. if we use 64 bits timestamps, it shouldn't be a problem to have timestamp in ns

Reply:
Let's use a custom timestamp representing the number of minutes since 1970-01-01T00:00:00 on 4 bytes. The remaining 8 bytes of the IV are generated randomly.

Reply:
> Let's use a custom timestamp representing the number of minutes since 1970-01-01T00:00:00 on 4 bytes. The remaining 8 bytes of the IV are generated randomly.

For context outcome of our private conversation:
- 32 bits seconds is not future proof enough.
- 64 bits seconds is overkill
- 64 bits nanoseconds is anoying on programing languages does have only have seconds precision included by default
- 64 bits is also anoying because that only leaves 4 bytes randomly generated in the IV, which makes collisions more likely than I like.
- 38 bits seconds + 7 random is heresy (not power of two words ?)

So here we go with 32 bits minutes precision.